ui/dfe.rb

```diff
@@ -0,0 +1,17 @@
+class Dfe
+  include Pages::Base
+
+  def initialize
+    page_classes = Dir.entries("#{File.dirname(__FILE__)}/pages")
+    page_classes.reject { |f| File.directory? f }.map { |x| x.gsub!('.rb', '') }
+    page_classes.delete_if { |file| file.match?(/base/) }
+
+    # NB: page names and file names must match!
+
+    page_classes.each do |result|
+      self.class.send(:define_method, result) do
+        "Pages::#{result.demodulize.camelize}".constantize.new
+      end
+    end
```

Comment:
This seems to be a really complicated way of requiring all the files in a directory. Why do you need to do it this way? Would it be easier just to do:
```ruby
Dir.glob("#{__dir__}/pages") { |f| require f }
```
Also constructing paths from the current file location can be problematic. It's usual practice to base your paths on the app root. So I'd suggest:
```ruby
Dir.glob(Rails.root.join("ui/pages/*.rb")) { |f| require f }
```

Reply:
OK - looks like I've lost the argument about metaprogramming.

However, I think it would be better to use `method_missing`. That will mean that the page class will only be loaded when it is needed, and it can also be easily cached.

Give me a short while and I'll put together some example code.

Reply:
I'm with you Rob, but the effort has been made, so I agree if our meta can be the best meta then even betta. Good call on `method_missing`.

Reply:
I'd suggest that this is a better option:
```ruby
  def pages
    @pages ||= {}
  end

  def method_missing(m, *args, &block)
     pages[m] ||= "Pages::#{m.to_s.demodulize.camelize}".constantize.new
  end
```
IMO the advantages are:
- Removes the File processing (I've tested locally and all the files needed seem to be loading without needing to be required individually)
- Only loads a page object when it is needed.
- Makes it easy to add caching (the `pages` object), so that only the first call on each `Dfe` instance, will initialize a new page instance. Remove this is a new object is needed for each call.
- Is simpler and more easily understandable code.

It fails nicely:
```
/../early-years-foundation-recovery/ui/dfe.rb:21:in `method_missing': undefined method `demodulize' for :foo:Symbol (NoMethodError)
```

Reply:
@reggieb `require` should be needed as the specs are gonna run without and automatic changes to LOADPATH courtesy of Rails

Reply:
@peterdavidhamilton Then this at the top of the file will fix that:
```ruby
Dir.glob(Rails.root.join("ui/pages/*.rb")) { |f| require f }
```

Reply:
@adeoke this gets my vote

Reply:
@reggieb @peterdavidhamilton Super. Glad you are both happy. Let me give it a go.

Reply:
done